lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/TaskManager.cs

```diff
@@ -0,0 +1,526 @@
+        /// </summary>
+        /// <param name="runningTask"></param>
+        internal void SetRunningTask(IRunningTask runningTask)
+        {
+            if (_runningTasks.ContainsKey(runningTask.Id))
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture, "The task [{0}] already running.", runningTask.Id);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            if (!_tasks.ContainsKey(runningTask.Id))
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture, "The task [{0}] doesn't exist.", runningTask.Id);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            _runningTasks.Add(runningTask.Id, runningTask);
+            UpdateState(runningTask.Id, TaskStateEvent.RunningTask);
+        }
+
+        /// <summary>
+        /// This method is called at the beginning of the recovery.
+        /// Clears the task collection, running task collection and resets the number of application error. 
+        /// </summary>
+        internal void Reset()
+        {
+            _tasks.Clear();
+            _runningTasks.Clear();
+            _numberOfAppError = 0;
+        }
+
+        /// <summary>
+        /// This method is called when receiving ICompletedTask event during task running or system shutting down.
+        /// Removes the task from running tasks
+        /// Changes the task state from RunningTask to CompletedTask
+        /// </summary>
+        /// <param name="taskId"></param>
+        internal void SetCompletedTask(string taskId)
+        {
+            RemoveRunningTask(taskId);
+            UpdateState(taskId, TaskStateEvent.CompletedTask);
+        }
+
+        /// <summary>
+        /// This method is called when receiving IFailedTask event during task submitting or running
+        /// Removes the task from running tasks if the task was running
+        /// Updates the task state to fail based on the error message in the failed task
+        /// </summary>
+        /// <param name="failedTask"></param>
+        internal void SetFailedRunningTask(IFailedTask failedTask)
+        {
+            RemoveRunningTask(failedTask.Id);
+            UpdateState(failedTask.Id, GetTaskErrorEvent(failedTask));
+        }
+
+        /// <summary>
+        /// This method is called when receiving IFailedTask event during system shutting down. 
+        /// If the task failed because it receives the close command from driver, update the task state to TaskClosedByDriver.
+        /// Task could fail by communication error or any other application or system error during this time, as long as it is not 
+        /// TaskFailedByEvaluatorFailure, update the task state based on the error received. 
+        /// </summary>
+        /// <param name="failedTask"></param>
+        internal void SetFailedTaskInShuttingDown(IFailedTask failedTask)
```

Comment:
If this is going to be the convention, please change all other functions that alter `TaskState` to begin with `Record` instead of `Set` as well.